crash when using extern structs with zero-sized fields

Zig version: `0.11.0-dev.3979+d78517f4f`

## Problem 1

```zig
const E = extern struct {
    foo: i32 = 0,
    bar: u0 = 0,
};

test {
    var e: E = .{};
    _ = e;
}
```

```
$ stage4/bin/zig  test test3.zig 
thread 3604715 panic: reached unreachable code
Analyzing test3.zig: test3.zig:test_0
      %10 = dbg_block_begin()
      %11 = dbg_stmt(2, 5)
      %12 = block_comptime({
        %13 = decl_val("E") token_offset:7:12 to :7:13
        %14 = break(%12, %13)
      }) node_offset:7:12 to :7:13
    > %15 = alloc_mut(%12) node_offset:7:5 to :7:19
      %16 = store_node(%15, @InternPool.Index.empty_struct) node_offset:7:16 to :7:19
      %17 = dbg_var_ptr(%15, "e")
      %18 = dbg_stmt(3, 5)
      %19 = load(%15) node_offset:8:9 to :8:10
      %20 = ensure_result_non_error(%19) node_offset:8:9 to :8:10
      %21 = dbg_block_end()
      %22 = restore_err_ret_index(%9, %4294967212)
      %23 = break(%9, @InternPool.Index.void_value)
    For full context, use the command
      zig ast-check -t test3.zig

  in test3.zig: test3.zig:test_0
    > %9 = block({%10..%23}) node_offset:6:6 to :6:7

/home/andy/Downloads/zig/lib/std/debug.zig:258:14: 0x5942e06 in assert (zig)
    if (!ok) unreachable; // assertion failure
             ^
/home/andy/Downloads/zig/lib/std/math.zig:1122:11: 0x5e27161 in isPowerOfTwo__anon_121360 (zig)
    assert(int > 0);
          ^
/home/andy/Downloads/zig/src/Sema.zig:24729:43: 0x5c3a21c in explainWhyTypeIsNotExtern (zig)
        .Int => if (!std.math.isPowerOfTwo(ty.intInfo(mod).bits)) {
                                          ^
/home/andy/Downloads/zig/src/Sema.zig:34548:51: 0x6197351 in semaStructFields (zig)
                try sema.explainWhyTypeIsNotExtern(msg, ty_src, field.ty, .struct_field);
                                                  ^
/home/andy/Downloads/zig/src/Sema.zig:34172:25: 0x5e21e71 in resolveTypeFieldsStruct (zig)
    try semaStructFields(sema.mod, struct_obj);
                        ^
/home/andy/Downloads/zig/src/Sema.zig:34119:53: 0x5e25826 in resolveTypeFields (zig)
                    try sema.resolveTypeFieldsStruct(ty, struct_obj);
                                                    ^
/home/andy/Downloads/zig/src/Sema.zig:24487:59: 0x668a2a6 in validateRunTimeType (zig)
            const resolved_ty = try sema.resolveTypeFields(ty);
                                                          ^
/home/andy/Downloads/zig/src/Sema.zig:24422:37: 0x65bee7f in validateVarType (zig)
    if (try sema.validateRunTimeType(var_ty, is_extern)) return;
                                    ^
/home/andy/Downloads/zig/src/Sema.zig:3792:29: 0x60a21a9 in zirAllocMut (zig)
    try sema.validateVarType(block, ty_src, var_ty, false);
                            ^
/home/andy/Downloads/zig/src/Sema.zig:923:66: 0x5e0a8aa in analyzeBodyInner (zig)
            .alloc_mut                    => try sema.zirAllocMut(block, inst),
                                                                 ^
/home/andy/Downloads/zig/src/Sema.zig:5680:34: 0x6691a89 in resolveBlockBody (zig)
        if (sema.analyzeBodyInner(child_block, body)) |_| {
                                 ^
/home/andy/Downloads/zig/src/Sema.zig:5663:33: 0x618962d in zirBlock (zig)
    return sema.resolveBlockBody(parent_block, src, &child_block, body, inst, &label.merges);
                                ^
/home/andy/Downloads/zig/src/Sema.zig:1485:49: 0x5e1d244 in analyzeBodyInner (zig)
                    break :blk try sema.zirBlock(block, inst, tags[inst] == .block_comptime);
                                                ^
/home/andy/Downloads/zig/src/Sema.zig:823:30: 0x6056298 in analyzeBody (zig)
    _ = sema.analyzeBodyInner(block, body) catch |err| switch (err) {
                             ^
/home/andy/Downloads/zig/src/Module.zig:5626:21: 0x5dee60d in analyzeFnBody (zig)
    sema.analyzeBody(&inner_block, fn_info.body) catch |err| switch (err) {
                    ^
/home/andy/Downloads/zig/src/Module.zig:4224:40: 0x5c2cfe9 in ensureFuncBodyAnalyzed (zig)
            var air = mod.analyzeFnBody(func_index, sema_arena) catch |err| switch (err) {
                                       ^
/home/andy/Downloads/zig/src/Compilation.zig:3146:42: 0x5c2afc9 in processOneJob (zig)
            module.ensureFuncBodyAnalyzed(func) catch |err| switch (err) {
                                         ^
/home/andy/Downloads/zig/src/Compilation.zig:3083:30: 0x5ac4a2e in performAllTheWork (zig)
            try processOneJob(comp, work_item, main_progress_node);
                             ^
/home/andy/Downloads/zig/src/Compilation.zig:2031:31: 0x5ac11f0 in update (zig)
    try comp.performAllTheWork(main_progress_node);
                              ^
/home/andy/Downloads/zig/src/main.zig:3862:24: 0x5aedeea in updateModule (zig)
        try comp.update(main_progress_node);
                       ^
/home/andy/Downloads/zig/src/main.zig:3297:17: 0x5995836 in buildOutputType (zig)
    updateModule(comp) catch |err| switch (err) {
                ^
/home/andy/Downloads/zig/src/main.zig:275:31: 0x5945064 in mainArgs (zig)
        return buildOutputType(gpa, arena, args, .zig_test);
                              ^
/home/andy/Downloads/zig/src/main.zig:213:20: 0x5942285 in main (zig)
    return mainArgs(gpa, arena, args);
                   ^
/home/andy/Downloads/zig/lib/std/start.zig:608:37: 0x5941d07 in main (zig)
            const result = root.main() catch |err| {
                                    ^
???:?:?: 0x7f06ede5424d in ??? (???)
???:?:?: 0x7fff586438df in ??? (???)
Aborted (core dumped)
```

## Problem 2

```zig
const E = extern struct {
    foo: i32 = 0,
    bar: void = {},
};

test {
    var e: E = .{};
    _ = e;
}
```

This should be legal, along with any other zero-sized type. However:

```zig
$ stage4/bin/zig  test test3.zig 
test3.zig:3:10: error: extern structs cannot contain fields of type 'void'
    bar: void = {},
         ^~~~
test3.zig:3:10: note: 'void' is a zero bit type; for C 'void' use 'anyopaque'
referenced by:
    E: test3.zig:1:18
    test_0: test3.zig:7:12
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```

## Workaround

```zig
const E = extern struct {
    foo: i32 = 0,
    bar: [0]u8 = .{},
};

test {
    var e: E = .{};
    _ = e;
}
```

Appears to work fine.

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -24590,7 +24590,7 @@ fn validateExternType(
         .ErrorSet,
         .Frame,
         => return false,
-        .Void => return position == .union_field or position == .ret_ty,
+        .Void => return position == .union_field or position == .ret_ty or position == .struct_field or position == .element,
         .NoReturn => return position == .ret_ty,
         .Opaque,
         .Bool,
@@ -24599,7 +24599,7 @@ fn validateExternType(
         => return true,
         .Pointer => return !(ty.isSlice(mod) or try sema.typeRequiresComptime(ty)),
         .Int => switch (ty.intInfo(mod).bits) {
-            8, 16, 32, 64, 128 => return true,
+            0, 8, 16, 32, 64, 128 => return true,
             else => return false,
         },
         .Fn => {
@@ -24620,11 +24620,11 @@ fn validateExternType(
             .Packed => {
                 const bit_size = try ty.bitSizeAdvanced(mod, sema);
                 switch (bit_size) {
-                    8, 16, 32, 64, 128 => return true,
+                    0, 8, 16, 32, 64, 128 => return true,
                     else => return false,
                 }
             },
-            .Auto => return false,
+            .Auto => return !(try sema.typeHasRuntimeBits(ty)),
         },
         .Array => {
             if (position == .ret_ty or position == .param_ty) return false;
@@ -24673,9 +24673,9 @@ fn explainWhyTypeIsNotExtern(
         .Void => try mod.errNoteNonLazy(src_loc, msg, "'void' is a zero bit type; for C 'void' use 'anyopaque'", .{}),
         .NoReturn => try mod.errNoteNonLazy(src_loc, msg, "'noreturn' is only allowed as a return type", .{}),
         .Int => if (!std.math.isPowerOfTwo(ty.intInfo(mod).bits)) {
-            try mod.errNoteNonLazy(src_loc, msg, "only integers with power of two bits are extern compatible", .{});
+            try mod.errNoteNonLazy(src_loc, msg, "only integers with 0 or power of two bits are extern compatible", .{});
         } else {
-            try mod.errNoteNonLazy(src_loc, msg, "only integers with 8, 16, 32, 64 and 128 bits are extern compatible", .{});
+            try mod.errNoteNonLazy(src_loc, msg, "only integers with 0, 8, 16, 32, 64 and 128 bits are extern compatible", .{});
         },
         .Fn => {
             if (position != .other) {

--- a/test/behavior/extern_struct_zero_size_fields.zig
+++ b/test/behavior/extern_struct_zero_size_fields.zig
@@ -1,0 +1,21 @@
+const E = enum(u0) {
+    the_only_possible_value,
+};
+
+const S = struct {};
+
+const T = extern struct {
+    foo: u0 = 0,
+    bar: void = {},
+    baz: struct {} = .{},
+    ayy: E = .the_only_possible_value,
+    arr: [0]u0 = .{},
+    matey: [128]void = [_]void{{}} ** 128,
+    running_out_of_ideas: packed struct {} = .{},
+    one_more: [256]S = [_]S{.{}} ** 256,
+};
+
+test {
+    var t: T = .{};
+    _ = t;
+}

--- a/test/cases/compile_errors/exported_enum_without_explicit_integer_tag_type.zig
+++ b/test/cases/compile_errors/exported_enum_without_explicit_integer_tag_type.zig
@@ -14,5 +14,5 @@ comptime {
 // :3:5: error: unable to export type 'type'
 // :7:5: error: unable to export type 'tmp.E'
 // :7:5: note: enum tag type 'u1' is not extern compatible
-// :7:5: note: only integers with 8, 16, 32, 64 and 128 bits are extern compatible
+// :7:5: note: only integers with 0, 8, 16, 32, 64 and 128 bits are extern compatible
 // :1:11: note: enum declared here

--- a/test/cases/compile_errors/extern_struct_with_extern-compatible_but_inferred_integer_tag_type.zig
+++ b/test/cases/compile_errors/extern_struct_with_extern-compatible_but_inferred_integer_tag_type.zig
@@ -43,5 +43,5 @@ export fn entry() void {
 //
 // :33:8: error: extern structs cannot contain fields of type 'tmp.E'
 // :33:8: note: enum tag type 'u9' is not extern compatible
-// :33:8: note: only integers with power of two bits are extern compatible
+// :33:8: note: only integers with 0 or power of two bits are extern compatible
 // :2:15: note: enum declared here

--- a/test/cases/compile_errors/extern_struct_with_non-extern-compatible_integer_tag_type.zig
+++ b/test/cases/compile_errors/extern_struct_with_non-extern-compatible_integer_tag_type.zig
@@ -13,5 +13,5 @@ export fn entry() void {
 //
 // :3:8: error: extern structs cannot contain fields of type 'tmp.E'
 // :3:8: note: enum tag type 'u31' is not extern compatible
-// :3:8: note: only integers with power of two bits are extern compatible
+// :3:8: note: only integers with 0 or power of two bits are extern compatible
 // :1:15: note: enum declared here

--- a/test/cases/compile_errors/extern_variable_has_non_extern_type.zig
+++ b/test/cases/compile_errors/extern_variable_has_non_extern_type.zig
@@ -8,4 +8,4 @@ pub export fn entry() void {
 // target=native
 //
 // :1:17: error: extern variable cannot have type 'u3'
-// :1:17: note: only integers with power of two bits are extern compatible
+// :1:17: note: only integers with 0 or power of two bits are extern compatible

--- a/test/cases/compile_errors/function_with_non-extern_non-packed_enum_parameter.zig
+++ b/test/cases/compile_errors/function_with_non-extern_non-packed_enum_parameter.zig
@@ -9,5 +9,5 @@ export fn entry(foo: Foo) void {
 //
 // :2:17: error: parameter of type 'tmp.Foo' not allowed in function with calling convention 'C'
 // :2:17: note: enum tag type 'u2' is not extern compatible
-// :2:17: note: only integers with 8, 16, 32, 64 and 128 bits are extern compatible
+// :2:17: note: only integers with 0, 8, 16, 32, 64 and 128 bits are extern compatible
 // :1:13: note: enum declared here

--- a/test/cases/compile_errors/variadic_arg_validation.zig
+++ b/test/cases/compile_errors/variadic_arg_validation.zig
@@ -24,6 +24,6 @@ pub export fn entry3() void {
 // :4:33: error: integer and float literals passed to variadic function must be casted to a fixed-size number type
 // :9:24: error: arrays must be passed by reference to variadic function
 // :13:24: error: cannot pass 'u48' to variadic function
-// :13:24: note: only integers with power of two bits are extern compatible
+// :13:24: note: only integers with 0 or power of two bits are extern compatible
 // :17:24: error: cannot pass 'void' to variadic function
 // :17:24: note: 'void' is a zero bit type; for C 'void' use 'anyopaque'

